### PR TITLE
perf_pmu.py: Adding test cases related to sysfs, registering events

### PR DESCRIPTION
**Pull request by nasastry**
Adding test cases,
What PMUs got registered
different kernel configs - CONFIG_PMU_SYSFS, CONFIG_PPC_RTAS
cpu events count
hv_24x7 events count
IMC events count
nest events CPU mask
Core events CPU mask

Signed-off-by: Nageswara R Sastry <rnsastry@linux.vnet.ibm.com>

### CODE REVIEW

**Comment by nasastry** [comment]
```
JOB ID     : 6420d9a4f339ab14ba8970868afc7c921f02d262
JOB LOG    : /home/OpTest/avocado-fvt-wrapper/results/job-2020-12-04T05.19-6420d9a/job.log
 (1/8) perf_pmu.py:PerfBasic.test_PMU_register: PASS (0.74 s)
 (2/8) perf_pmu.py:PerfBasic.test_config_PMU_sysfs: PASS (1.96 s)
 (3/8) perf_pmu.py:PerfBasic.test_config_PPC_RTAS: PASS (2.08 s)
 (4/8) perf_pmu.py:PerfBasic.test_cpu_event_count: PASS (0.77 s)
 (5/8) perf_pmu.py:PerfBasic.test_hv_24x7_event_count: PASS (0.65 s)
 (6/8) perf_pmu.py:PerfBasic.test_imc_event_count: SKIP: This test is for PowerNV
 (7/8) perf_pmu.py:PerfBasic.test_nest_cpumask: SKIP: This test is for PowerNV
 (8/8) perf_pmu.py:PerfBasic.test_core_cpumask: SKIP: This test is for PowerNV
RESULTS    : PASS 5 | ERROR 0 | FAIL 0 | SKIP 3 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/OpTest/avocado-fvt-wrapper/results/job-2020-12-04T05.19-6420d9a/results.html
JOB TIME   : 9.33 s
```

**Comment by nasastry** [comment]
Though it gives LGTM alerts for the variables power8, 9 and 10
these variables are used using 'eval' in the following line.
131 self._check_file_existence(eval(self.model), os.listdir(sysfs_file))

File: perf/perf_pmu.py [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+                                       ignore_status=True)
+        # TBD: Check for any error should be the issue
+        if b'Permission denied' in output:
+            self.fail("Unable to read mmcr* files as super user.")
+
+        self._create_temp_user()
+        output = process.system_output("su - test_pmu -c 'cat %smmcr*'"
+                                       % sysfs_file, shell=True,
+                                       allow_output_check='combined',
+                                       ignore_status=True)
```
Comment (harish-24) [inline perf/perf_pmu.py:146]: `.decode()` can be used here

File: perf/perf_pmu.py [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+        # Check for any missing files according to the model
+        self._check_file_existence(eval(self.model), os.listdir(sysfs_file))
+
+        # allow_output_check='combined' helps in combining stdout, stderr
+        output = process.system_output("cat %smmcr*" % sysfs_file,
+                                       sudo=True, shell=True,
+                                       allow_output_check='combined',
+                                       ignore_status=True)
+        # TBD: Check for any error should be the issue
+        if b'Permission denied' in output:
```
Comment (harish-24) [inline perf/perf_pmu.py:139]: Will this error not be part of `stderr`?. `system_output` only returns `stdout`.

File: perf/perf_pmu.py [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+            self.fail("Unable to read mmcr* files as super user.")
+
+        self._create_temp_user()
+        output = process.system_output("su - test_pmu -c 'cat %smmcr*'"
+                                       % sysfs_file, shell=True,
+                                       allow_output_check='combined',
+                                       ignore_status=True)
+        self._remove_temp_user()
+        if b'Permission denied' not in output:
```
Comment (harish-24) [inline perf/perf_pmu.py:148]: here too, make sure if `stderr` is needed

File: perf/perf_pmu.py [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+
+        power8 = ['mmcr0', 'mmcr1', 'mmcra']
+        power9 = ['mmcr0', 'mmcr1', 'mmcra']
+        power10 = ['mmcr0', 'mmcr1', 'mmcr3', 'mmcra']
+
+        # Check for any missing files according to the model
+        self._check_file_existence(eval(self.model), os.listdir(sysfs_file))
+
+        # allow_output_check='combined' helps in combining stdout, stderr
+        output = process.system_output("cat %smmcr*" % sysfs_file,
```
Comment (harish-24) [inline perf/perf_pmu.py:134]: why not use `genio.read_file` for reading the file content?

File: perf/perf_pmu.py [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+        output = process.system_output("su - test_pmu -c 'cat %smmcr*'"
+                                       % sysfs_file, shell=True,
+                                       allow_output_check='combined',
+                                       ignore_status=True)
+        self._remove_temp_user()
+        if b'Permission denied' not in output:
+            self.fail("Able to read mmcr* files as normal user.")
+
+    def _verify_lscpu_sysfs(self, filename, value_to_verify):
+        output = process.system_output("cat %s" % filename, sudo=True,
```
Comment (harish-24) [inline perf/perf_pmu.py:152]: `genio` module can be used here

File: perf/perf_pmu.py [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+    def test_PMU_register(self):
+        # This function tests whether performance monitor hardware support
+        # registered or not. If not found any registered messages in dmesg
+        # output this test will fail.
+        output = process.system_output("dmesg").decode()
+        for line in output.split('\n'):
+            if 'performance monitor hardware support registered' in line:
+                matchFound = re.search(r"\] (.*) performance", line)
+                if matchFound:
+                    self.pmu_list.append(matchFound.group(1))
```
Comment (harish-24) [inline perf/perf_pmu.py:94]: `dmesg.collect_errors_dmesg` can be used

File: perf/perf_pmu.py [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+                                      % (user_cmd, sysfs_file), self.pchips)
+        self._verify_lscpu_sysfs_test("%s 'cat %scoresperchip'"
+                                      % (user_cmd, sysfs_file), self.pcorechips)
+        self._remove_temp_user()
+
+        if self.pchips == "1" and self.psockets == "1" and self.pcorechips == "1":
+            output = process.system_output("dmesg").decode()
+            for line in output.split('\n'):
+                if 'rtas error: Error calling get-system-parameter' in line:
+                    self.fail("RTAS error occured")
```
Comment (harish-24) [inline perf/perf_pmu.py:196]: `dmesg.collect_errors_dmesg` can be used

File: perf/perf_pmu.py [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+
+    def _check_count(self, event_type):
+        base_dir = "/sys/bus/event_source/devices/"
+        if not os.path.isdir(base_dir):
+            self.fail("sysfs events folder not found")
+
+        if not os.path.isdir(base_dir + event_type):
+            self.fail("sysfs %s folder not found" % event_type)
+
+        sys_fs_events = os.listdir(base_dir + event_type + '/events/')
```
Comment (harish-24) [inline perf/perf_pmu.py:206]: `os.path.join` can be used in few places in this function

File: perf/perf_pmu.py [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+    # TBD: collect only cpu, hv_24x7, imc
+    def _sysfs_pmu_entries(self, pmu_type):
+        sysfs_pmu = os.listdir('/sys/devices/')
+        sysfs_pmu_list = []
+        for entry in sysfs_pmu:
+            if pmu_type in entry:
+                sysfs_pmu_list.append(entry)
+        if not len(sysfs_pmu_list) > 0:
+            self.fail("%s PMUs not found in sysfs" % pmu_type)
+        else:
```
Comment (harish-24) [inline perf/perf_pmu.py:240]:
```
    if not len(sysfs_pmu_list):
        self.fail()
    return ...
```

File: perf/perf_pmu.py [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+    def _sysfs_pmu_cpumask(self, pmu_list):
+        failed_list = []
+        for pmu in pmu_list:
+            cpumask_file = '/sys/devices/' + pmu + '/cpumask'
+            pmu_cpumask = open(cpumask_file).readlines()
+            if not len(pmu_cpumask) > 0:
+                failed_list.append(pmu_cpumask)
+            self.log.info("%s contains cpumask = %s" % (cpumask_file,
+                                                        pmu_cpumask))
+        if len(failed_list) > 0:
```
Comment (harish-24) [inline perf/perf_pmu.py:252]: `if len(failed_list):`

File: perf/perf_pmu.py (hunk 2 above)
Comment (nasastry) [reply perf/perf_pmu.py:139]: I am combining the stderr and stdout.

File: perf/perf_pmu.py (hunk 3 above)
Comment (nasastry) [reply perf/perf_pmu.py:148]: I am combining the stderr and stdout.

File: perf/perf_pmu.py (hunk 5 above)
Comment (nasastry) [reply perf/perf_pmu.py:152]: I want to see the 'permission denied error' if it occurs in-case of any non-sufficient privileges.

**Comment by nasastry** [comment]
Latest code change run results:

RHEL
```
JOB ID     : e94aa5a0c78465e82b7631be273c6c1c39a92c15
JOB LOG    : /home/avocado-fvt-wrapper/results/job-2021-01-25T05.30-e94aa5a/job.log
 (1/9) perf_pmu.py:PerfBasic.test_PMU_register: PASS (1.20 s)
 (2/9) perf_pmu.py:PerfBasic.test_config_PMU_sysfs: PASS (1.95 s)
 (3/9) perf_pmu.py:PerfBasic.test_config_PPC_RTAS: PASS (1.91 s)
 (4/9) perf_pmu.py:PerfBasic.test_cpu_event_count: PASS (0.64 s)
 (5/9) perf_pmu.py:PerfBasic.test_hv_24x7_event_count: PASS (0.60 s)
 (6/9) perf_pmu.py:PerfBasic.test_imc_event_count: SKIP: This test is for PowerNV
 (7/9) perf_pmu.py:PerfBasic.test_thread_event_count: SKIP: This test is for PowerNV
 (8/9) perf_pmu.py:PerfBasic.test_nest_cpumask: SKIP: This test is for PowerNV
 (9/9) perf_pmu.py:PerfBasic.test_core_cpumask: SKIP: This test is for PowerNV
RESULTS    : PASS 5 | ERROR 0 | FAIL 0 | SKIP 4 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/avocado-fvt-wrapper/results/job-2021-01-25T05.30-e94aa5a/results.html
JOB TIME   : 9.93 s
```

SLES
```
JOB ID     : 657dea6fbe7d27d3804b6b8e4ead30eeb5df487a
JOB LOG    : /home/avocado-fvt-wrapper/results/job-2021-01-25T16.00-657dea6/job.log
 (1/9) perf_pmu.py:PerfBasic.test_PMU_register: PASS (1.52 s)
 (2/9) perf_pmu.py:PerfBasic.test_config_PMU_sysfs: FAIL: For power10, Not found files are ['mmcr0', 'mmcr1', 'mmcr3', 'mmcra'] (0.56 s)
 (3/9) perf_pmu.py:PerfBasic.test_config_PPC_RTAS: PASS (0.80 s)
 (4/9) perf_pmu.py:PerfBasic.test_cpu_event_count: PASS (0.49 s)
 (5/9) perf_pmu.py:PerfBasic.test_hv_24x7_event_count: PASS (0.50 s)
 (6/9) perf_pmu.py:PerfBasic.test_imc_event_count: SKIP: This test is for PowerNV
 (7/9) perf_pmu.py:PerfBasic.test_thread_event_count: SKIP: This test is for PowerNV
 (8/9) perf_pmu.py:PerfBasic.test_nest_cpumask: SKIP: This test is for PowerNV
 (9/9) perf_pmu.py:PerfBasic.test_core_cpumask: SKIP: This test is for PowerNV
RESULTS    : PASS 4 | ERROR 0 | FAIL 1 | SKIP 4 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/avocado-fvt-wrapper/results/job-2021-01-25T16.00-657dea6/results.html
JOB TIME   : 6.94 s
```

**Comment by nasastry** [comment]
@harish-24 Requested changed were made. Please have a look. Thank you.

File: perf/perf_pmu.py (hunk 5 above)
Comment (harish-24) [reply perf/perf_pmu.py:146]: Prefer using the `os.stat` and `stat` module of python to achieve that. And eventually the output can still be obtained with the genio module.

File: perf/perf_pmu.py (hunk 4 above)
Comment (harish-24) [reply perf/perf_pmu.py:129]: same here

File: perf/perf_pmu.py (hunk 8 above)
Comment (harish-24) [reply perf/perf_pmu.py:199]: Any reason for leaving this out?

File: perf/perf_pmu.py (hunk 9 above)
Comment (harish-24) [reply perf/perf_pmu.py:233]: else is not required here since the code is already past the condition and hence must always return a valid value

**Review by harish-24 — CHANGES_REQUESTED** [review]

**Comment by nasastry** [comment]
```
JOB ID     : 50d48f35a1b3018348c402c91a9685b152d70828
JOB LOG    : /home/avocado-fvt-wrapper/results/job-2021-02-02T00.24-50d48f3/job.log
 (1/9) perf_pmu.py:PerfBasic.test_PMU_register: PASS (0.78 s)
 (2/9) perf_pmu.py:PerfBasic.test_config_PMU_sysfs: PASS (2.92 s)
 (3/9) perf_pmu.py:PerfBasic.test_config_PPC_RTAS: PASS (2.96 s)
 (4/9) perf_pmu.py:PerfBasic.test_cpu_event_count: PASS (0.60 s)
 (5/9) perf_pmu.py:PerfBasic.test_hv_24x7_event_count: PASS (0.59 s)
 (6/9) perf_pmu.py:PerfBasic.test_imc_event_count: SKIP: This test is for PowerNV
 (7/9) perf_pmu.py:PerfBasic.test_thread_event_count: SKIP: This test is for PowerNV
 (8/9) perf_pmu.py:PerfBasic.test_nest_cpumask: SKIP: This test is for PowerNV
 (9/9) perf_pmu.py:PerfBasic.test_core_cpumask: SKIP: This test is for PowerNV
RESULTS    : PASS 5 | ERROR 0 | FAIL 0 | SKIP 4 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/avocado-fvt-wrapper/results/job-2021-02-02T00.24-50d48f3/results.html
JOB TIME   : 11.22 s
```

**Comment by nasastry** [comment]
Handled undefined variable:

```
JOB ID     : 22e7351dbb75abece1097746cdae4fde4e3ee245
JOB LOG    : /home/avocado-fvt-wrapper/results/job-2021-02-02T01.08-22e7351/job.log
 (1/9) perf_pmu.py:PerfBasic.test_PMU_register: PASS (0.74 s)
 (2/9) perf_pmu.py:PerfBasic.test_config_PMU_sysfs: PASS (2.80 s)
 (3/9) perf_pmu.py:PerfBasic.test_config_PPC_RTAS: PASS (3.36 s)
 (4/9) perf_pmu.py:PerfBasic.test_cpu_event_count: PASS (0.62 s)
 (5/9) perf_pmu.py:PerfBasic.test_hv_24x7_event_count: PASS (0.57 s)
 (6/9) perf_pmu.py:PerfBasic.test_imc_event_count: SKIP: This test is for PowerNV
 (7/9) perf_pmu.py:PerfBasic.test_thread_event_count: SKIP: This test is for PowerNV
 (8/9) perf_pmu.py:PerfBasic.test_nest_cpumask: SKIP: This test is for PowerNV
 (9/9) perf_pmu.py:PerfBasic.test_core_cpumask: SKIP: This test is for PowerNV
RESULTS    : PASS 5 | ERROR 0 | FAIL 0 | SKIP 4 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/avocado-fvt-wrapper/results/job-2021-02-02T01.08-22e7351/results.html
JOB TIME   : 11.44 s
```

**Review by harish-24 — APPROVED** [review]
LGTM

**Pull request merged**